### PR TITLE
Support for Listen interval based ps

### PR DESCRIPTION
**Pull request by ajayparida**
@sachinthegreen @krish2718 @rado17

### CODE REVIEW

File: include/zephyr/net/wifi_mgmt.h [hunk 1]
```diff
@@ -274,6 +289,14 @@ enum wifi_twt_sleep_state {
 	WIFI_TWT_STATE_AWAKE = 1,
 };
 
+struct wifi_listen_interval_params {
+	unsigned short listen_interval;
+};
+
+struct wifi_ps_wake_interval_params {
```
Comment (krish2718) [inline include/zephyr/net/wifi_mgmt.h:296]: This is not an interval, its wakeup mode

File: include/zephyr/net/wifi_mgmt.h [hunk 2]
```diff
@@ -274,6 +289,14 @@ enum wifi_twt_sleep_state {
 	WIFI_TWT_STATE_AWAKE = 1,
 };
 
+struct wifi_listen_interval_params {
+	unsigned short listen_interval;
```
Comment (krish2718) [inline include/zephyr/net/wifi_mgmt.h:293]: As discussed earlier, having a OP for each param is an overkill, can you please group these parameters and simplify the ops?

File: subsys/net/l2/wifi/wifi_mgmt.c [hunk 3]
```diff
@@ -343,3 +353,47 @@ void wifi_mgmt_raise_twt_sleep_state(struct net_if *iface,
...
+	struct wifi_listen_interval_params *params = data;
+	struct wifi_iface_status info = { 0 };
+
+	if (off_api == NULL || off_api->set_listen_interval == NULL) {
+		return -ENOTSUP;
+	}
+
+	if (net_mgmt(NET_REQUEST_WIFI_IFACE_STATUS, iface, &info,
+		     sizeof(struct wifi_iface_status))) {
+		return -ENOTSUP;
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_mgmt.c:372]: Shouldn't this be EIO or something?

File: subsys/net/l2/wifi/wifi_shell.c [hunk 4]
```diff
@@ -900,9 +961,25 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
...
+		      0),
+	SHELL_CMD_ARG(ps_listen_interval,
+		      NULL,
+		      "<val> - Listen interval in the range of <3-65535>",
+		      cmd_wifi_listen_interval,
+		      2,
+		      0),
+	SHELL_CMD_ARG(ps_wake_interval,
+		     NULL,
+		     "<dtim> : Set ps wake interval to Dtim\n"
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_shell.c:978]:
```suggestion
		     "<dtim> : Set PS wake up period to DTIM interval\n"
```

File: subsys/net/l2/wifi/wifi_shell.c [hunk 5]
```diff
@@ -900,9 +961,25 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
...
+		     "<listen_interval> : Set ps wake interval to listen interval",
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_shell.c:979]:
```suggestion
		     "<listen_interval> : Set PS wake up period to listen interval",
```

**Review by krish2718 — CHANGES_REQUESTED** [review]

File: include/zephyr/net/wifi_mgmt.h [hunk 6]
```diff
@@ -189,17 +177,21 @@ struct wifi_iface_status {
...
 	enum wifi_ps_mode mode;
-};
-
-struct wifi_ps_timeout_params {
+	char num_twt_flows;
 	int timeout_ms;
-};
+	unsigned short listen_interval;
+	enum wifi_ps_wakeup_mode wakeup_mode;
+	int ps_flag;
```
Comment (krish2718) [inline include/zephyr/net/wifi_mgmt.h:193]: This should be named `oper` or `subcmd`, my earlier review comment to aggregate commands was to simplify the configuration, but this `subcmd` basically re-introduces the sub-commands but using a single structure. Also, the division of parameters doesn't looks correct:

* `num_twt_flows` is in `wifi_ps_params` but `twt_flows` is in `wifi_ps_config`
* `wifi_ps_config` and `wifi_ps_params` are synonymous
* `enabled` and `timeout_ms` in `wifi_ps_params` are dynamic parameters i.e., can be called at any time, but rest of them are only pre-association (mode, listen_interval)

I would suggest below split:

```
/* Static - should be configured before association and cannot be changed during lifetime */
struct wifi_ps_static_config {
   enum wifi_ps_mode mode;
   unsigned short listen_interval;
};

/* Dynamic PS config - Can be send any time, before or after association, and can be changed anytime */
struct wifi_ps_dynamic_config {
  enum wifi_ps enabled;
  unsigned int timeout_ms;
  enum wifi_ps_wakeup_mode wakeup_mode;   
  char num_twt_flows;
  struct wifi_twt_flow_info twt_flows[WIFI_MAX_TWT_FLOWS];
};
```

We could separate out TWT if that makes handling easier.

File: subsys/net/l2/wifi/wifi_shell.c [hunk 7]
```diff
@@ -843,6 +847,65 @@ static int cmd_wifi_reg_domain(const struct shell *sh, size_t argc,
...
+	int err = 0;
+
+	context.sh = sh;
+
+	if (!parse_number(sh, &interval, argv[1], 3, 65535)) {
+		return -EINVAL;
+	}
+
+	params.listen_interval = interval;
+	params.ps_flag = WIFI_PS_PARAM_LISTEN_INTERVAL;
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_shell.c:864]: This `subcmd` setting should be in wifi_mgmt.c as its a direct match b/w `NET_REQUEST_WIFI_LISTEN_INTERVAL` and is redundant to set both

File: subsys/net/l2/wifi/wifi_shell.c [hunk 8]
```diff
@@ -843,6 +847,65 @@ static int cmd_wifi_reg_domain(const struct shell *sh, size_t argc,
...
+
+	if (net_mgmt(NET_REQUEST_WIFI_LISTEN_INTERVAL, iface, &params, sizeof(params))) {
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_shell.c:866]: `NET_REQUEST_WIFI_LISTEN_INTERVAL` where is this defined, I cannot seem to find it anywhere?

**Review by krish2718 — CHANGES_REQUESTED** [review]
Overall at the design level we have two options: 

1. Treat each parameter as a separate configuration with its own structure, shell command and an API
2. Group parameters as suggested pre-assoc and anytime

The latest changes seem to do a hybrid, in which case we can simply make it an union of parameters as we only seem to allow individual parameter configuration. 

We really need to fix the design to avoid code churn in review.

@rlubos  any thoughts?

**Comment by krish2718** [comment]
> > Overall at the design level we have two options:
> > 
> > 1. Treat each parameter as a separate configuration with its own structure, shell command and an API
> > 2. Group parameters as suggested pre-assoc and anytime
> > 
> > The latest changes seem to do a hybrid, in which case we can simply make it an union of parameters as we only seem to allow individual parameter configuration.
> > We really need to fix the design to avoid code churn in review.
> > @rlubos any thoughts?
> 
> I'm not sure if I get the full context (it would really help if you started to provide the overall goal of the PR in the description) but when it comes to radio configuration, the device we have in 15.4 driver seems to work fine.

This PR is adding many OPs to configure individual params, so, was proposing a way to aggregate these for easier understanding and maintenance.

> 
> Instead of multiplying APIs/grouping configs we have single API function to set a config (https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/net/ieee802154_radio.h#L356) and each config is represented with an enum value. The config value is provided as an union (https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/net/ieee802154_radio.h#L231).

This looks reasonable, and is close to what the current changes in the PR do, so, we can proceed with that, thanks.

File: subsys/net/l2/wifi/wifi_shell.c [hunk 9]
```diff
@@ -843,6 +847,65 @@ static int cmd_wifi_reg_domain(const struct shell *sh, size_t argc,
 	return 0;
 }
 
+static int cmd_wifi_listen_interval(const struct shell *sh, size_t argc, char *argv[])
+{
+	struct net_if *iface = net_if_get_default();
+	struct wifi_ps_params params = { 0 };
+	long interval = 10;
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_shell.c:854]: Do we need a random default here?

File: subsys/net/l2/wifi/wifi_shell.c [hunk 10]
```diff
@@ -843,6 +847,65 @@ static int cmd_wifi_reg_domain(const struct shell *sh, size_t argc,
...
+		shell_fprintf(sh, SHELL_WARNING, "Invalid argument\n");
+		shell_fprintf(sh, SHELL_INFO,
+			      "Valid argument : <dtim> / <listen_interval>\n");
+		return -ENOEXEC;
+	}
+
+	params.type = WIFI_PS_PARAM_WAKEUP_MODE;
+
+	if (net_mgmt(NET_REQUEST_WIFI_PS, iface, &params, sizeof(params))) {
+		shell_fprintf(sh, SHELL_WARNING, "Setting ps wake period to %s failed\n",
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_shell.c:898]:
```suggestion
		shell_fprintf(sh, SHELL_WARNING, "Setting ps wake up mode to %s failed\n",
```

File: subsys/net/l2/wifi/wifi_shell.c [hunk 11]
```diff
@@ -453,12 +453,21 @@ static int cmd_wifi_ps(const struct shell *sh, size_t argc, char *argv[])
...
+				wifi_ps2str[config.ps_params.enabled]);
 		if (config.enabled) {
 			shell_fprintf(sh, SHELL_NORMAL, "PS mode: %s\n",
 					wifi_ps_mode2str[config.mode]);
 		}
 
+		shell_fprintf(sh, SHELL_NORMAL, "PS listen_interval: %d\n",
+				config.ps_params.listen_interval);
+
+		shell_fprintf(sh, SHELL_NORMAL, "PS wake up period: %s\n",
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_shell.c:465]:
```suggestion
		shell_fprintf(sh, SHELL_NORMAL, "PS wake up mode: %s\n",
```

File: subsys/net/l2/wifi/wifi_shell.c [hunk 12]
```diff
@@ -900,9 +969,25 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
...
+		      0),
+	SHELL_CMD_ARG(ps_listen_interval,
+		      NULL,
+		      "<val> - Listen interval in the range of <3-65535>",
+		      cmd_wifi_listen_interval,
+		      2,
+		      0),
+	SHELL_CMD_ARG(ps_wakeup_mode,
+		     NULL,
+		     "<dtim> : Set PS wake up period to DTIM interval\n"
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_shell.c:986]:
```suggestion
		     "<dtim> : Set PS wake up mode to DTIM interval\n"
```

**Review by krish2718 — CHANGES_REQUESTED** [review]

**Review by rado17 — DISMISSED** [review]

File: include/zephyr/net/wifi.h [hunk 13]
```diff
@@ -14,6 +14,13 @@
 
 #define WIFI_COUNTRY_CODE_LEN 2
 
+/* As per hardware design for power efiicienecy
+ * minimum listen interval period should be
+ * 3 DTIM period.
+ */
+#define WIFI_LISTEN_INTERVAL_MIN 3
```
Comment (krish2718) [inline include/zephyr/net/wifi.h:21]: As this is generic, we should follow protocol limits, any HW specific limits should be enforced by the underlying driver.

**Review by krish2718 — CHANGES_REQUESTED** [review]

**Review by rado17 — DISMISSED** [review]

**Review by krish2718 — DISMISSED** [review]

File: subsys/net/l2/wifi/wifi_shell.c [hunk 14]
```diff
@@ -912,9 +1001,25 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
...
-		  cmd_wifi_ps_timeout),
+	SHELL_CMD_ARG(ps_timeout,
+		      NULL,
+		      "<val> - Power save inactivity timer(in ms)",
+		      cmd_wifi_ps_timeout,
+		      2,
+		      0),
+	SHELL_CMD_ARG(ps_listen_interval,
+		      NULL,
+		      "<val> - Listen interval in the range of <3-65535>",
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_shell.c:1012]: This still shows 3

File: subsys/net/l2/wifi/wifi_shell.c [hunk 15]
```diff
@@ -912,9 +1001,25 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
 		"-f: Force to use this regulatory hint over any other regulatory hints\n"
 		"Note: This may cause regulatory compliance issues, use it at your own risk.",
 		cmd_wifi_reg_domain),
-	SHELL_CMD(ps_timeout, NULL,
-		  "Configure Wi-Fi power save inactivity timer(in ms)",
-		  cmd_wifi_ps_timeout),
+	SHELL_CMD_ARG(ps_timeout,
+		      NULL,
+		      "<val> - Power save inactivity timer(in ms)",
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_shell.c:1006]: Use either PS or Power save but consistently.

File: subsys/net/l2/wifi/wifi_shell.c [hunk 16]
```diff
@@ -855,6 +865,79 @@ static int cmd_wifi_reg_domain(const struct shell *sh, size_t argc,
...
+		shell_fprintf(sh, SHELL_INFO,
+			      "Valid argument : <dtim> / <listen_interval>\n");
+		return -ENOEXEC;
+	}
+
+	params.type = WIFI_PS_PARAM_WAKEUP_MODE;
+
+	if (net_mgmt(NET_REQUEST_WIFI_PS, iface, &params, sizeof(params))) {
+		shell_fprintf(sh, SHELL_WARNING,
+			      "Setting ps wake up mode to %s failed..Reason :%s\n",
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_shell.c:929]: PS

**Review by krish2718 — DISMISSED** [review]
Nits.

File: subsys/net/l2/wifi/wifi_shell.c (hunk 15 above)
Comment (ajayparida) [reply subsys/net/l2/wifi/wifi_shell.c]: used PS

**Review by krish2718 — APPROVED** [review]

**Review by sachinthegreen — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Comment by ajayparida** [comment]
@jukkar

**Review by sachinthegreen — APPROVED** [review]

**Review by krish2718 — APPROVED** [review]

**Review by jukkar — APPROVED** [review]
LGTM.

I noticed that the wifi shell uses this pattern
```
struct net_if *iface = net_if_get_default();
```
This can be a bit problematic as it probably only works reliably if there is one network interface in the system.

**Comment by krish2718** [comment]
> LGTM.
> 
> I noticed that the wifi shell uses this pattern
> 
> ```
> struct net_if *iface = net_if_get_default();
> ```
> 
> This can be a bit problematic as it probably only works reliably if there is one network interface in the system.

Yes, definitely, it crashed for multi-VIF (netusb, 802.15.4). I have been meaning to address this, in downstream we have worked around by always setting Wi-Fi as default interface.

**Comment by jukkar** [comment]
> I have been meaning to address this, in downstream we have worked around by always setting Wi-Fi as default interface

That is certainly possible but I would also recommend that user could give the network interface index as a parameter in various wifi commands (for example "-i <iface index>" or similar)

**Pull request merged**